cloudinit/persistence.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (C) 2020 Canonical Ltd.
+#
+# Author: Daniel Watkins <oddbloke@ubuntu.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+
+class CloudInitPickleMixin:
+    """Scaffolding for versioning of pickles.
+
+    This class implements ``__getstate__`` and ``__setstate__`` to provide
+    lightweight versioning of the pickles that are generated for classes which
+    use it.  Versioning is done at the class level.
+
+    The current version of a class's pickle should be set in the class variable
+    ``_ci_pkl_version``, as an int.  If not overriden, it will default to 0.
+
+    On unpickle, the object's state will be restored and then
+    ``self._unpickle`` is called with the version of the stored pickle as the
+    only argument: this is where classes should implement any deserialization
+    fixes they require.  (If the stored pickle has no version, 0 is passed.)
+    """
+
+    _ci_pkl_version = 0
+
+    def __getstate__(self):
+        """Persist instance state, adding a pickle version attribute.
+
+        This adds a ``_ci_pkl_version`` attribute to ``self.__dict__`` and
+        returns that for serialisation.  The attribute is stripped out in
+        ``__setstate__`` on unpickle.
+
+        The value of ``_ci_pkl_version`` is ``type(self)._ci_pkl_version``.
+        """
+        state = self.__dict__.copy()
+        state["_ci_pkl_version"] = type(self)._ci_pkl_version
```

Comment:
Nit: Is there a reason for `type(self)`? `self` on it's own should point to the right thing.

Reply:
It's maybe overly defensive, but it avoids `_ci_pkl_version` ever being read from instance state, and so protects us against, e.g., someone writing `self._ci_pkl_version = ...`. We want to use the version that is in the on-disk class definition and this protects against one avenue where that could be modified. (Of course, someone could still do `type(self)._ci_pkl_version = ...` so it isn't a perfect guard.)

```py
In [3]: class Example(CloudInitPickleMixin): 
   ...:     _ci_pkl_version = 2 
   ...:     def __init__(self): 
   ...:         self._ci_pkl_version = 3 

In [4]: Example()._ci_pkl_version                                                                                                                                                                                                                                                         
Out[4]: 3

In [5]: type(Example())._ci_pkl_version                                                                                                                                                                                                                                                   
Out[5]: 2

In [6]: type(Example())._ci_pkl_version = 3                                                                                                                                                                                                                                               

In [7]: Example()._ci_pkl_version                                                                                                                                                                                                                                                         
Out[7]: 3
```

Reply:
Or, put another way: `_ci_pkl_version` is a property of the type, not of instances of the type; as such, we look it up on the type directly.